meta/client.go

```diff
@@ -0,0 +1,215 @@
+package meta
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/XiaoMi/pegasus-go-client/idl/base"
+	"github.com/sirupsen/logrus"
+	"log"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/XiaoMi/pegasus-go-client/session"
+	"github.com/bluele/gcache"
+	"github.com/go-zookeeper/zk"
+)
+
+// TODO(jiashuo) store config file
+var zkAddrs = []string{""}
+var zkTimeOut = 1000000000 // unit ns, equal 1s
+var zkRoot = "/pegasus-cluster"
+var zkWatcherCount = 1024
+
+var globalClusterManager *ClusterManager
+
+type ClusterManager struct {
+	Mut    sync.RWMutex
+	ZkConn *zk.Conn
+	// table->TableInfoWatcher
+	Tables gcache.Cache
+	// metaAddrs->metaManager
+	Metas map[string]*session.MetaManager
+}
+
+type Context struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+type TableInfoWatcher struct {
+	tableName   string
+	clusterName string
+	metaAddrs   string
+	event       <-chan zk.Event
+	context     Context
+}
+
+// TODO(jishuo1) change log module
+func initClusterManager() {
+	zkConn, _, err := zk.Connect(zkAddrs, time.Duration(zkTimeOut))
+	if err != nil {
+		panic(fmt.Errorf("connect to %s failed: %s", zkAddrs, err))
+	}
+
+	tables := gcache.New(zkWatcherCount).LRU().EvictedFunc(func(key interface{}, value interface{}) {
+		value.(*TableInfoWatcher).context.cancel()
+		logrus.Warnf("table cache has exceeded %d and remove the oldest table[%s]", zkWatcherCount, key.(string))
+	}).Build() // TODO(jiashuo1) consider set expire time
+	globalClusterManager = &ClusterManager{
+		ZkConn: zkConn,
+		Tables: tables,
+		Metas:  make(map[string]*session.MetaManager),
+	}
+}
+
+func (m *ClusterManager) getMetaConnector(table string) (*session.MetaManager, error) {
+	var metaConnector *session.MetaManager
+
+	tableInfo, err := globalClusterManager.Tables.Get(table)
+	if err == nil {
+		metaConnector = globalClusterManager.Metas[tableInfo.(*TableInfoWatcher).metaAddrs]
+		if metaConnector != nil {
+			return metaConnector, nil
+		}
+	}
+
+	log.Printf("[%s] can't get cluster info from local cache, try fetch from zk.", table)
+	globalClusterManager.Mut.Lock()
+	tableInfo, err = globalClusterManager.Tables.Get(table)
+	if err != nil {
+		tableInfo, err = m.getTableInfo(table)
+		if err != nil {
+			logrus.Errorf("get table[%s] info failed: %s", table, err)
+			return nil, err
+		}
+		err = globalClusterManager.Tables.Set(table, tableInfo)
+		if err != nil {
+			log.Printf("[%s] cluster info update local cache failed: %s", table, err)
```

Comment:
```suggestion
			log.Panicf("[%s] cluster info update local cache failed: %s", table, err)
```
we don't expect gcache.set (https://github.com/bluele/gcache/blob/master/lru.go#L29) could fail. As a LRU cache, it should never fail on set.

Reply:
I only think the `Cache` is optional, we should tolerate the few error, In this case, our service still keep runing but crush